File: src/main/java/fr/ayfri/minecraft_art/TextureManager.java
package fr.ayfri.minecraft_art;

import processing.core.PImage;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.util.Objects;

public class TextureManager {
	private final Main sketch;
	public static final String[] EXCEPTIONS = { "top", "bottom", "brewing_stand", "cauldron", "hopper", "slab", "shulker", "anvil", "wheat" };
	
	// todo Exceptions gérées par le JSON
	public TextureManager(final Main sketch) {
		this.sketch = sketch;
	}
	
	public void init() throws IOException, URISyntaxException {
		final File blocks = ResourceUtils.getLocalFile("./blocks");
		final File[] files = Objects.requireNonNull(blocks).listFiles();
		
		assert files != null;
		for (final File blockFile : files) {
			if (blockFile.getName().endsWith(".png") && Utils.listNotContainsItem(EXCEPTIONS, blockFile.getName())) {
				final PImage image = getImage(blockFile);
				final int averageColor = Utils.getAverageColor(image);
				boolean cancelRegister = false;
				image.loadPixels();
				
				if (image.height != 16 || image.width != 16) {
					continue;
				}
				
				for (final int pixel : image.pixels) {
					final float colorDistance = Utils.getColorDistance(sketch.g, pixel, averageColor);
					if (colorDistance > 60 || sketch.alpha(pixel) < 255) {
						cancelRegister = true;
						break;
					}
				}
				
				if (cancelRegister) {
					continue;
				}
				
				Main.BLOCKS.put(blockFile.getName(), image);
				Main.COLORS_OF_BLOCKS.put(blockFile.getName(), averageColor);
			}
		}
	}
	
	public PImage getImage(final File file) throws FileNotFoundException {
		if (file.exists()) {
			return sketch.loadImage(file.getPath());
		} else {
			throw new FileNotFoundException("File " + file.getPath() + " not found.");
		}
	}
}

File: src/main/kotlin/io/github/ayfri/minecraft_art/TextureManager.kt
package io.github.ayfri.minecraft_art

import processing.core.PImage
import java.io.File
import java.io.FileNotFoundException

// todo Exceptions gérées par le JSON
data class TextureManager(private val sketch: Main) {
	/** Registers every flat, opaque 16x16 texture of the local `blocks` directory into [Blocks]. */
	fun init() {
		val files = getLocalDirectory("blocks")?.listFiles() ?: return

		for (blockFile in files) {
			val name = blockFile.name
			if (!name.endsWith(".png") || EXCEPTIONS.any { it in name }) continue

			val image = getImage(blockFile)
			image.loadPixels()
			if (image.width != 16 || image.height != 16) continue

			val averageColor = getAverageColor(image)
			/** A texture is kept only when fully opaque and flat enough, so the block reads as a single solid color once scaled down. */
			val isFlatAndOpaque = image.pixels.all {
				getColorDistance(sketch.g, it, averageColor) <= 60 && sketch.alpha(it) >= 255
			}
			if (!isFlatAndOpaque) continue

			Blocks.register(name, image, averageColor)
		}
	}

	fun getImage(file: File): PImage {
		if (!file.exists()) throw FileNotFoundException("File ${file.path} not found.")
		return sketch.loadImage(file.path)
	}

	companion object {
		val EXCEPTIONS = listOf("top", "bottom", "brewing_stand", "cauldron", "hopper", "slab", "shulker", "anvil", "wheat")
	}
}
